### PR TITLE
Safe hide and reveal

### PR DESCRIPTION
**Pull request by krufab**
What does this implement/fix? Explain your changes.
---------------------------------------------------
Extends git-secret functionality making it more reliable in case of merging secrets from different branches which might have evolved independently, thus generating conflicts and differences which have to be manually solved.

Does this close any currently open issues?
------------------------------------------
This PR fixes issue [#328](https://github.com/sobolevn/git-secret/issues/328).

### CODE REVIEW

File: tests/_test_base.bash [hunk 1]
```diff
@@ -53,6 +53,26 @@ TEST_DEFAULT_FILENAME="space file" # has spaces
...
+function debug_output {
+  # both variables are inherited from the test
+  echo "$output" | sed "s/^/# '$BATS_TEST_DESCRIPTION' output:/" >&3
+}
+
+# Prints the content of lines variable.
+# NOTE: empty strings (lines) are not preserved, and are not printed out
+# in this function. Nevertheless, this is the actual content of $lines
+# which can be used in tests like [ "$lines[1] = "something" ].
+# Usage: debug_lines
```
[the remark this answers is not in the document]
Comment (krufab) [reply tests/_test_base.bash]: Well, it's a handy function I used while writing tests, to check the content of the output. And yes, $lines, $output and $BATS_TEST_DESCRIPTION are bats parameters

**Comment by joshrabinowitz** [comment]
@krufab This PR is really impressive and nicely done, thank you. 

I like seeing the complete picture of the implementation, but I think it's too big a change for one PR. 

What do you think about submitting these features as a series of PRs, each PR adding just one feature with tests? Then we can discuss the features and implementation more easily as we go.

**Comment by krufab** [comment]
@joshrabinowitz Thanks!
Well, I could start with a PR for the git secret changes module, then with the hide one and for last, git secret reveal. I'll be slightly busy at work this week, but let's see if i can find some time.

**Comment by krufab** [comment]
Hi! I've created a new PR (I didn't want to roll back the code of this branch). It's here: [PR#358](https://github.com/sobolevn/git-secret/pull/358)

**Comment by joshrabinowitz** [comment]
Since we are not planning to merge this PR, and instead plan to move towards #328 with a series of smaller PRs, I'm closing this PR. 
This PR/branch will remain available for checkout and investigation as needed.

**Pull request closed without merging**